### PR TITLE
Change the decoding routines to return errors instead of panicing.

### LINKED ISSUE #2671: sql: improve error handling when decoding table keys
`sql.decodeTableKey` currently blindly assumes that the encoded data matches the expected column type. If the expectation doesn't match we can get panics in the decoding routines. We're not currently and haven't historically seen panics from this code, but it is probably worthwhile to be defensive. 

We could change the decoding routines to return an error instead of panicing, though that seems onerous to the callers. We could add some sort of verification routine that can be called first, though that also places a burden on the callers to remember to call that routine. On the other hand, the decoding routines are not called from that many places (at least in the `sql` package) so perhaps any worry about additional burden is misplaced.

### PR DESCRIPTION
**Pull request by petermattis**
See #2671.

### CODE REVIEW

**Comment by petermattis** [comment]
The `encoding.MustDecode*` routines are intended to be a stop-gap. Attempting to make sure every path to the encoding routines returned an error felt burdensome for this PR. There are only a handful of usages of the `encoding.MustDecode*` routines:

```
~/Development/go/src/github.com/cockroachdb/cockroach pmattis/decoding-error git grep encoding.MustDecode
config/config.go:       _, id64 := encoding.MustDecodeUvarint(remaining)
keys/keys.go:   _, rangeID := encoding.MustDecodeUvarint(b)
keys/keys.go:   b, startKey = encoding.MustDecodeBytes(b, nil)
keys/keys.go:           _, k = encoding.MustDecodeBytes(k, nil)
storage/engine/mvcc.go: tsBytes, key := encoding.MustDecodeBytes(encodedKey, nil)
storage/engine/mvcc.go: tsBytes, walltime = encoding.MustDecodeUint64Decreasing(tsBytes)
storage/engine/mvcc.go: tsBytes, logical = encoding.MustDecodeUint32Decreasing(tsBytes)
ts/keys.go:     remainder, name = encoding.MustDecodeBytes(remainder, nil)
ts/keys.go:     remainder, resolutionInt = encoding.MustDecodeVarint(remainder)
ts/keys.go:     remainder, timeslot = encoding.MustDecodeVarint(remainder)
```

The usage in `storage/engine/mvcc.go` looks to be the most irritating to fix.

File: util/encoding/float.go [hunk 1]
```diff
@@ -95,45 +95,45 @@ func EncodeFloat(b []byte, f float64) []byte {
...
-		return buf[idx+1:], makeFloatFromMandE(false, e, m, tmp)
+		return buf[idx+1:], makeFloatFromMandE(false, e, m, tmp), nil
 	case buf[0] == floatPosSmall:
 		// Positive small.
 		e, m := decodeSmallNumber(false, buf[:idx+1], tmp)
-		return buf[idx+1:], makeFloatFromMandE(false, e, m, tmp)
+		return buf[idx+1:], makeFloatFromMandE(false, e, m, tmp), nil
 	default:
-		panic(fmt.Sprintf("unknown prefix of the encoded byte slice: %q", buf))
+		return nil, 0, fmt.Errorf("unknown prefix of the encoded byte slice: %q", buf)
```
Comment (tamird) [inline util/encoding/float.go]: util.Errorf?

File: util/encoding/encoding.go [hunk 2]
```diff
@@ -534,14 +587,20 @@ func EncodeTime(b []byte, t time.Time) []byte {
 
 // DecodeTime converts an encoded time into a UTC time.Time type and
 // returns the rest of the buffer.
-func DecodeTime(b []byte) ([]byte, time.Time) {
+func DecodeTime(b []byte) ([]byte, time.Time, error) {
 	if PeekType(b) != Time {
-		panic("did not find marker")
+		return nil, time.Time{}, fmt.Errorf("did not find marker")
```
Comment (tamird) [inline util/encoding/encoding.go]: util?

**Comment by tamird** [comment]
LGTM modulo fmt/util.Errorf confusion

**Comment by tamird** [comment]
Oh, and I'd leave #2671 open.

File: util/encoding/float.go (hunk 1 above)
Comment (petermattis) [inline util/encoding/float.go]: Yeah. Done everywhere.

Comment (tamird) [inline util/encoding/float.go]: heh, not here

Comment (petermattis) [inline util/encoding/float.go]: Just making sure you were paying attention.

**Pull request merged**